src/cider/include/cider/CiderAllocator.h

```diff
@@ -111,4 +113,110 @@ class AlignAllocator : public CiderAllocator {
 
   std::shared_ptr<CiderAllocator> parent_;
 };
+
+class AllocatedData {
+ public:
+  AllocatedData(CiderAllocator* allocator, int8_t* pointer, size_t allocated_size)
+      : allocator_(allocator), pointer_(pointer), allocated_size_(allocated_size) {}
+  ~AllocatedData() { reset(); }
+
+  int8_t* get() { return pointer_; }
+  const int8_t* get() const { return pointer_; }
+  size_t getSize() const { return allocated_size_; }
+  void reset() {
+    if (!pointer_) {
+      return;
+    }
+    allocator_->deallocate(pointer_, allocated_size_);
+    pointer_ = nullptr;
+  }
+
+ private:
+  CiderAllocator* allocator_;
+  int8_t* pointer_;
+  size_t allocated_size_;
+};
+
+struct ArenaChunk {
+  ArenaChunk(CiderAllocator* allocator, size_t size)
+      : current_position(0)
+      , maximum_size(size)
+      , prev(nullptr)
+      , data(allocator, allocator->allocate(size), size) {}
+  ~ArenaChunk() {
+    if (next) {
+      auto current_next = std::move(next);
+      while (current_next) {
+        current_next = std::move(current_next->next);
+      }
+    }
+  }
+
+  AllocatedData data;
+  size_t current_position;
+  size_t maximum_size;
+  std::unique_ptr<ArenaChunk> next;
+  ArenaChunk* prev;
+};
+
+class CiderArenaAllocator : public CiderAllocator {
+  static constexpr const size_t ARENA_ALLOCATOR_INITIAL_CAPACITY = 2048;
+
+ public:
+  CiderArenaAllocator(
+      std::shared_ptr<CiderAllocator> parent = std::make_shared<CiderDefaultAllocator>(),
+      size_t initial_capacity = ARENA_ALLOCATOR_INITIAL_CAPACITY)
+      : parent_(parent) {
+    head_ = nullptr;
+    tail_ = nullptr;
+    current_arena_capacity_ = initial_capacity;
+  }
+
+  int8_t* allocate(size_t len) final {
+    if (!head_ || head_->current_position + len > head_->maximum_size) {
```

Comment:
does this allocation algorithm refer existing ones? If left size is not enough for new allocation request, new arena chunk will be created with 2X size initialized. But what about the left space of current arena chunk? Will it cause big memory fragmentation?

Reply:
yes, this arena allocator referred DuckDB impl. I think it will not cause memory fragment for small string. For large string it may have such issue.